src/ophyd_async/core/_detector.py

```diff
@@ -331,19 +354,29 @@ async def complete(self):
                 + (self._trigger_info.deadtime or 0)
             )
         ):
+            target = (
+                self._trigger_info.number[self._current_number_index]
+                if isinstance(self._trigger_info.number, list)
+                else self._trigger_info.number
+            )
             yield WatcherUpdate(
                 name=self.name,
                 current=index,
                 initial=self._initial_frame,
-                target=self._trigger_info.number,
+                target=target,
                 unit="",
                 precision=0,
                 time_elapsed=time.monotonic() - self._fly_start,
             )
-            if index >= self._trigger_info.number:
-                break
-        if self._iterations_completed == self._trigger_info.iteration:
-            self._iterations_completed = 0
+            if isinstance(self._trigger_info.number, list):
+                if index >= self._trigger_info.number[self._current_number_index]:
+                    self._current_number_index += 1
+                    break
+            else:
+                if index >= self._trigger_info.number:
+                    break
+        if self._frames_collected == self._total_frames:
+            self._frames_collected = 0
```

Comment:
I think if we're breaking from an async generator then it doesn't necessary close? We probably need something like

``` python
indices_written = self.writer.observe_indices_written(
    self._trigger_info.frame_timeout or (
        DEFAULT_TIMEOUT
        + (self._trigger_info.livetime or 0)
        + (self._trigger_info.deadtime or 0)
    )
)
try:
    async for index in indices_written:
        # .... DO YOUR STUFF AND BREAK
finally:
    await indices_written.aclose()
```